### PR TITLE
treesitter: Better configuration of queries (set files explicitly, deferred evaluation in set_query)

### PR DESCRIPTION
**Pull request by theHamsta**
Allow functions in set_query for deferred evaluation. This allows to determine the query string just when needed. Before one would have to put `set_query` in a file specific config file in order to not steal precious start-up time (including the parsing of the queries). This is for users who want to have precise control on how the queries are constructed.

However, most user just want to keep the regular mechanism and just want to replace a few files. So when we have `get_query/set_query` we could also have `get_query_files/set_query_files`.

Usage examples
```lua
          require("vim.treesitter.query").set_query_files("python", "folds", {"/home/stephan/foo.scm"})

          -- All runtime files except those from nvim-treesitter
          require("vim.treesitter.query").set_query_files("python", "folds", vim.tbl_filter(function(f) return not f:find('nvim-treesitter', 1, true) end, require"vim.treesitter.query".get_runtime_files("python", "folds"))
          -- Or wrap it into a fn to not read from disk during config time
          require("vim.treesitter.query").set_query_files("python", "folds", function() vim.tbl_filter(function(f) return not f:find('nvim-treesitter', 1, true) end, require"vim.treesitter.query".get_runtime_files("python", "folds") end)
```

This PR is not squashed in docs some polishing is still missing (e.g. tests) but I want to hear your opinions first!

This isn't a ideal solution. My previous PR was to just allow functions in set `set_query` to run them deferred. i.e. parsing and search for runtime files would only occur when it's needed. Then, nvim-treesitter could provide some functions to filter-out it's query files. The first draft is https://github.com/neovim/neovim/pull/14932/commits/4fdf1190c0d2ffbb5a5fe0be6ee3eee57c833f30

Addresses https://github.com/nvim-treesitter/nvim-treesitter/issues/1441

### CODE REVIEW

File: runtime/lua/vim/treesitter/query.lua [hunk 1]
```diff
@@ -99,9 +99,23 @@ end
```
Comment (theHamsta) [inline runtime/lua/vim/treesitter/query.lua:99]: a the moment `get_query_files` does not regard `explicit_query_files` but it probably should

**Comment by theHamsta** [comment]
@gjeusel  I didn't find a good solution for this problem. But something like this would enable to set the desired files easily or implement in `nvim-treesitter` to filter out it's own queries.

File: runtime/lua/vim/treesitter/query.lua [hunk 2]
```diff
@@ -114,9 +128,28 @@ local explicit_queries = setmetatable({}, {
 ---
 --- @param lang string: The language to use for the query
 --- @param query_name string: The name of the query (i.e. "highlights")
---- @param text string: The query text (unparsed).
+--- @param text string|function: The query text (unparsed) or function returning the text.
 function M.set_query(lang, query_name, text)
-  explicit_queries[lang][query_name] = M.parse_query(lang, text)
+  explicit_queries[lang][query_name] = text
```
Comment (theHamsta) [inline runtime/lua/vim/treesitter/query.lua:133]: get_query parses the queries every time anyways when they come from the runtime files. get_query needs to be cached anyways as reloading the queries was often too slow for most users. But we could also keep it

**Comment by justinmk** [comment]
`vim.treesitter.config` is discussed in https://github.com/neovim/neovim/pull/15260

Or ideally, we gain dict options and this can be configured via `:set foo = { ... }`

**Comment by clason** [comment]
@justinmk No, this is a different topic and will not be solved by `vim.treesitter.config`; this issue is about dynamic queries (so related to allowing `set_query` to _extend_ instead of replace queries from `rtp`, or adding an `#if-query?` predicated (which was rejected 😞 )).

**Pull request closed without merging**